core/src/main/java/org/apache/calcite/schema/lookup/CachedLookup.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.schema.lookup;
+
+import org.apache.calcite.util.LazyReference;
+import org.apache.calcite.util.NameMap;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Set;
+
+/**
+ * This class can be used to make a snapshot of a lookups.
+ *
+ * @param <T> Element Type
+ */
+public class CachedLookup<T> implements Lookup<T> {
+
+  private final Lookup<T> delegate;
+  private LazyReference<Lookup<T>> cachedDelegate = new LazyReference<>();
```

Comment:
I am assuming this is the "snapshot".
I would call it "snapshot" or "delegateSnapshot".
Is this class mixing two functionalities? Caching and snapshotting look to me to be separate behaviors.

Reply:
I renamed it to `SnapshotLookup`